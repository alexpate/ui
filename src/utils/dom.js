/**
 * Gets the current scroll position of the document element
 * http://stackoverflow.com/questions/3464876/javascript-get-window-x-y-position-for-scroll
 * @returns {{top: number, left: number}}
 */
export function getDocumentScroll() {
  const doc = document.documentElement;
  const left = (window.pageXOffset || doc.scrollLeft) - (doc.clientLeft || 0);
  const top = (window.pageYOffset || doc.scrollTop) - (doc.clientTop || 0);
  return {top, left};
}

/**
 * Calculate the fully absolute (to body top/left) position of an element
 * @param el
 * @returns {{top: number, left: number}}
 */
export function getFixedOffset(el) {
  const rect = !!el ? el.getBoundingClientRect() : {top: 0, left: 0};
  const scroll = getDocumentScroll();
  return {
    top: rect.top + scroll.top,
    left: rect.left + scroll.left,
  };
}

/**
 * Remove a given element from the DOM
 * @param {Element} el
 */
export function removeNode(el) {
  if (el && el.parentNode) {
    el.parentNode.removeChild(el);
  }
}

export function outerHeight(el) {
  const style = window.getComputedStyle(el);
  return (
    (el.offsetHeight || 0) +
    parseInt(style.marginTop || 0, 10) +
    parseInt(style.marginBottom || 0, 10)
  );
}

export function outerWidth(el) {
  const style = window.getComputedStyle(el);
  return (
    (el.offsetWidth || 0) +
    parseInt(style.marginLeft || 0, 10) +
    parseInt(style.marginRight || 0, 10)
  );
}

/**
 * Return the selection range from an element as start & end row/column vectors
 * NOTE: Rows and columns are 1-indexed
 *
 * @param el
 * @returns {[[startRow, startColumn], [endRow, endColumn], [startIndex, endIndex]}
 */
export function getSelectionRange(el) {
  const value = el.value;

  // Calculate selction start position
  const rowstoStart = value.substr(0, el.selectionStart).split('\n');
  const startRow = rowstoStart.length;
  const startColumn = rowstoStart[rowstoStart.length - 1].length + 1;

  // Calculate selection end position
  const rowsToEnd = value.substr(0, el.selectionEnd).split('\n');
  const endRow = rowsToEnd.length;
  const endColumn = rowsToEnd[rowsToEnd.length - 1].length + 1;

  // Return as vector range array
  return [
    [startRow, startColumn],
    [endRow, endColumn],
    [el.selectionStart, el.selectionEnd],
  ];
}

/**
 * Shift columns of given selection range by delta
 */
export function shiftSelectionColumns(selectionRange, delta) {
  return [
    [selectionRange[0][0], selectionRange[0][1] + delta],
    [selectionRange[1][0], selectionRange[1][1] + delta],
    [selectionRange[2][0] + delta, selectionRange[2][1] + delta],
  ];
}

/**
 * Takes a selection range array (as seen above in getSelectionRange)
 * and sets it on an input or textarea element
 */
export function setSelectionRange(el, selectionRange) {
  el.setSelectionRange(selectionRange[2][0], selectionRange[2][1]);
}

/**
 * Get the browsers scrollbar width
 * http://stackoverflow.com/questions/13382516/getting-scroll-bar-width-using-javascript
 */
export function getScrollbarWidth() {
  const outer = document.createElement('div');
  outer.style.visibility = 'hidden';
  outer.style.width = '100px';
  outer.style.msOverflowStyle = 'scrollbar'; // needed for WinJS apps

  document.body.appendChild(outer);

  const widthNoScroll = outer.offsetWidth;
  // force scrollbars
  outer.style.overflow = 'scroll';

  // add innerdiv
  const inner = document.createElement('div');
  inner.style.width = '100%';
  outer.appendChild(inner);

  const widthWithScroll = inner.offsetWidth;

  // remove divs
  outer.parentNode.removeChild(outer);

  return widthNoScroll - widthWithScroll;
}

export function addClass(el, className) {
  if (el.classList) {
    el.classList.add(className);
  } else {
    el.className = `${el.className} ${className}`; // eslint-disable-line no-param-reassign
  }
}

export function getParentBoundingClientRect(elem) {
  const htmlElem = elem instanceof HTMLElement ? elem : elem.parentElement; // eslint-disable-line no-undef
  return htmlElem.offsetParent.getBoundingClientRect();
}

export function createEvent(name, data = {}) {
  try {
    return new window.CustomEvent(name, {detail: data});
  } catch (e) {
    const event = document.createEvent('CustomEvent');
    event.initCustomEvent(name, true, true, data);
    return event;
  }
}
