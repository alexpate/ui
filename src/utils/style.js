import {isArray, isNumber, pickBy, omitBy} from 'lodash';

/**
 * Lookup to determine what props we include in style
 *
 * For now this is limited purely to style rules dealing with
 * layout and sizing (cur. box & text layout).
 */
const STYLE_WHITELIST = {
  alignItems: true,
  alignContent: true,
  alignSelf: true,
  borderWidth: true,
  borderTopWidth: true,
  borderRightWidth: true,
  borderBottomWidth: true,
  borderLeftWidth: true,
  display: true,
  cursor: true,
  flex: true,
  flexBasis: true,
  flexDirection: true,
  flexFlow: true,
  flexGrow: true,
  flexShrink: true,
  flexWrap: true,
  justifyContent: true,
  order: true,
  padding: true,
  paddingTop: true,
  paddingRight: true,
  paddingBottom: true,
  paddingLeft: true,
  margin: true,
  marginTop: true,
  marginRight: true,
  marginBottom: true,
  marginLeft: true,
  position: true,
  top: true,
  right: true,
  bottom: true,
  left: true,
  width: true,
  height: true,
  minWidth: true,
  minHeight: true,
  maxWidth: true,
  maxHeight: true,
  float: true,
  zIndex: true,
  overflow: true,
  overflowX: true,
  overflowY: true,
  tableLayout: true,
  verticalAlign: true,
  textAlign: true,
  textAlignLast: true,
  textIndent: true,
  textOverflow: true,
  textTransform: true,
  whiteSpace: true,
  wordBreak: true,
  wordSpacing: true,
  wordWrap: true,
  lineHeight: true,
  boxSizing: true,
};

const MAPPED_SPACING_PROPS = {
  margin: 'm',
  marginTop: 'mt',
  marginRight: 'mr',
  marginBottom: 'mb',
  marginLeft: 'ml',
  padding: 'p',
  paddingTop: 'pt',
  paddingRight: 'pr',
  paddingBottom: 'pb',
  paddingLeft: 'pl',
};

/*
 * Flexbox stlyle overrides for Safari 8
 * Safari 8 detection is performed in advance
 */
const VENDOR_STYLERS =
  window.navigator.userAgent.indexOf('Safari/') !== -1 &&
  window.navigator.userAgent.indexOf('Version/8') !== -1
    ? {
        display: (key, value) => ({
          key,
          value: (value.indexOf('flex') > -1 ? '-webkit-' : '') + value,
        }),
        alignContent: (key, value) => ({key: 'WebkitAlignContent', value}),
        alignItems: (key, value) => ({key: 'WebkitAlignItems', value}),
        alignSelf: (key, value) => ({key: 'WebkitAlignSelf', value}),
        justifyContent: (key, value) => ({key: 'WebkitJustifyContent', value}),
        order: (key, value) => ({key: 'WebkitOrder', value}),
        flexDirection: (key, value) => ({key: 'WebkitFlexDirection', value}),
        flexWrap: (key, value) => ({key: 'WebkitFlexWrap', value}),
        flexFlow: (key, value) => ({key: 'WebkitFlexFlow', value}),
        flex: (key, value) => ({key: 'WebkitFlex', value}),
        flexBasis: (key, value) => ({key: 'WebkitFlexBasis', value}),
        flexShrink: (key, value) => ({key: 'WebkitFlexShrink', value}),
        flexGrow: (key, value) => ({key: 'WebkitFlexGrow', value}),
      }
    : {};

/*
  * Lookup to determine which styles can be given a vector of values
  * Eg. Pass {padding: [0, 5, 5, 0]}
  */

const VECTOR_STYLES = {
  padding: true,
  margin: true,
};

/** Parse a specific style */
function parseStyle(name, value) {
  if (VECTOR_STYLES[name] && isArray(value)) {
    return value.map(v => (v ? (isNumber(v) ? `${v}px` : v) : 0)).join(' ');
  }

  return value;
}

/** Pull out styles from props */
export function parseStyleProps(rawProps, shouldAddVendors = true) {
  const props = {};
  const style = {};

  for (const key in rawProps) {
    if ({}.hasOwnProperty.call(rawProps, key)) {
      if (STYLE_WHITELIST[key]) {
        const parsedStyleValue = parseStyle(key, rawProps[key]);
        if (shouldAddVendors && !!VENDOR_STYLERS[key]) {
          const pair = VENDOR_STYLERS[key](key, parsedStyleValue);
          style[pair.key] = pair.value;
        } else {
          style[key] = parsedStyleValue;
        }
      } else {
        props[key] = rawProps[key];
      }
    }
  }

  return {props, style};
}

/**
 * @summary
 * Pull out styles from props, as well as map our
 * custom spacing prop names to ones that our layout
 * library expects:
 * 
 * @example: marginLeft => ml
 */
export function parseStyleAndSpacingProps(rawProps) {
  const spacingProps = {};
  const otherProps = {};

  for (const key in rawProps) {
    if ({}.hasOwnProperty.call(rawProps, key)) {
      if (MAPPED_SPACING_PROPS[key]) {
        spacingProps[MAPPED_SPACING_PROPS[key]] = parseStyle(
          key,
          rawProps[key]
        );
      } else {
        otherProps[key] = rawProps[key];
      }
    }
  }

  const {props, style} = parseStyleProps(otherProps, false);

  return {props, spacingProps, style};
}

/** Filter keys on an object to those from our STYLE_WHITELIST */
export function pickStyles(obj) {
  return pickBy(obj, (v, key) => STYLE_WHITELIST[key]);
}

export function omitStyles(obj) {
  return omitBy(obj, (v, key) => STYLE_WHITELIST[key]);
}
