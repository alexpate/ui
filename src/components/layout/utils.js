import React, {PureComponent} from 'react';
import {parseStyleProps, cleanProps} from '../../utils/style';
import cx from 'classnames';

export const makePrimitive = (name, DefaultComponent, defaultStyle) => {
  class Primitive extends PureComponent {
    render() {
      const {
        component: Component = DefaultComponent,
        elRef,
        style: propStyle,
        className,
        ...otherProps
      } = this.props;

      const {props, style} = parseStyleProps({
        ...otherProps,
      });

      const cleanedProps = cleanProps(props);

      return (
        <Component
          ref={elRef}
          className={className ? cx(className) : null}
          css={{...defaultStyle, ...style}}
          {...cleanedProps}
          style={{...propStyle}}
        />
      );
    }
  }

  Primitive.displayName = `Primitive${name}`;
  return Primitive;
};
