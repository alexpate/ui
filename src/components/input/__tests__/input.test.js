/* @flow */
/* eslint-env mocha */
import React from 'react';
import {shallow} from 'enzyme';
import expect from 'expect';
import {inline} from 'utils/string';

import Input from 'components/input';

describe('components/input', () => {
  describe('Input (default)', () => {
    const newInput = (props) => {
      const wrapper = shallow(
        <Input
          {...props}
        />
      );
      const input = wrapper.find('input');
      const inputStyle = input.props().style;
      return {inputStyle};
    };

    it(inline(`
      should have a bottom border by default
    `), () => {
      const {inputStyle} = newInput();
      expect(inputStyle.borderBottomWidth).toBe('1px');
    });

    it(inline(`
      should have no bottom border if \`borderless\` is true
    `), () => {
      const {inputStyle} = newInput({borderless: true});
      expect(inputStyle.borderBottomWidth).toBe('0');
    });
  });
});
