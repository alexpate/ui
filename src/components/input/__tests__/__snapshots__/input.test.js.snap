// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render shallow component ok 1`] = `
<div
  className="css-13nj03s1 css-11q3kj7"
  style={Object {}}
>
  <input
    className="css-kus6en css-13nj03s2"
    defaultValue={undefined}
    name={undefined}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder={undefined}
    readOnly={undefined}
    size="medium"
    style={
      Object {
        "width": undefined,
      }
    }
    value={undefined}
  />
</div>
`;
